### PR TITLE
[TF-TRT] Changes to the MatMul and Binary Converters

### PR DESCRIPTION
**Pull request by pranavm-nvidia**
This change consists of 2 major parts:
1. Changes to the MatMul converter to use IMatrixMultiplyLayer instead of IFullyConnectedLayer where it is advantageous. The latter requires inputs to be at least rank 3, whereas the former does not. The IFullyConnectedLayer is only used when either int8 mode is specified (MM does not support int8) or when the input is already rank >= 3. 
2. Changes to the Binary converter to use the IElementWiseLayer instead of IScaleLayer. Once again, the latter requires inputs to be at least rank 3, whereas the former does not. Replacing it greatly simplifies the conversion logic and broadens support for some ops.

### CODE REVIEW

**Comment by rthadur** [comment]
@pranavm-nvidia thank you for your contribution , can you please sign CLA

**Comment by pranavm-nvidia** [comment]
I've signed the CLA

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 1]
```diff
@@ -3219,32 +2964,68 @@ Status ConvertBinary(OpConverterParams* params) {
...
+  };
+  auto op_pair = ops.find(node_def.op());
+  if (op_pair == ops.end()) {
+    return errors::Unimplemented("Binary op ", node_def.op(),
+                                 " not supported at: ", node_def.name());
   }
-  return status;
+
+  nvinfer1::Dims broadcasted_dims_l, broadcasted_dims_r;
+  Status status = params->converter->GetTrtBroadcastShape(
```
Comment (trevor-m) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: `params->converter` is `nullptr` during validation. This hasn't caused any problems yet because `GetTrtBroadcastShape` doesn't try to access any member variables at the moment. @aaroey should we make `GetTrtBroadcastShape` a standalone function?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc [hunk 2]
```diff
@@ -2149,7 +1967,26 @@ TEST_F(OpConverterTest, ConvertBinary) {
   TestBinaryTensorOpTensor<ops::Minimum, DT_FLOAT>(this);
   TestBinaryTensorOpTensor<ops::Maximum, DT_FLOAT>(this);
   TestBinaryTensorOpTensor<ops::Pow, DT_FLOAT>(this);
-
+  // Test with operand R = Weights
```
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc]: nit: it would be good to use templates to avoid duplicating these calls for different function/dtypes. I think we can do something like https://stackoverflow.com/a/4697356/4691962.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 3]
```diff
@@ -3976,44 +3757,102 @@ Status ConvertGather(OpConverterParams* params) {
   return Status::OK();
 }
 
-Status ConvertMatMulHelper(OpConverterParams* params,
-                           TRT_TensorOrWeights tensor_input,
-                           TRT_ShapedWeights weights_raw, bool transpose_weight,
-                           string node_name) {
-  if (!tensor_input.is_tensor()) {
-    return errors::InvalidArgument("Input 0 expects tensor");
+Status ConvertFCHelper(OpConverterParams* params, nvinfer1::ITensor* tensor_a,
```
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Would you please split the MatMul change to another PR? It's self-contained and putting it here makes it hard to review this PR.
Thanks.

**Review by aaroey — CHANGES_REQUESTED** [review]
Thanks a lot for the improvements!

Please split the MatMul change to another PR, since it's self-contained and putting it here makes it hard to review this PR.

**Comment by pranavm-nvidia** [comment]
I've created a new PR for the MatMul converter changes. I will remove them from this PR in the next commit (so this will only contain the binary converter/broadcasting changes)

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc [hunk 4]
```diff
@@ -977,43 +977,37 @@ TEST_F(ConverterTest, GetTrtBroadcastShape) {
...
   symmetric_test({1, 1, 1}, {2, 1, 1, 1}, kIsTensor, kIsNotTensor, {}, {},
-                 error::INVALID_ARGUMENT, "Infeasible broadcast scheme",
-                 /*operand_1_batch_size=*/2);
-  symmetric_test({1, 1, 1}, {1, 1, 1, 1, 1}, kIsTensor, kIsNotTensor, {}, {},
                  error::INVALID_ARGUMENT,
-                 "Broadcasting beyond batch dimension is not supported "
-                 "(tensor #dims 4 vs broadcast #dims 5)");
+                 "Cannot broadcast weights with non-trivial batch dimension",
+                 /*operand_1_batch_size=*/2);
+  symmetric_test({1, 1, 1}, {1, 1, 1, 1, 1}, kIsTensor, kIsNotTensor,
```
[the remark this answers is not in the document]
Comment (trevor-m) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc]: I don't think this is possible in TRT since all tensors must have the same batch size.

**Comment by aaroey** [comment]
Closing this w.r.t #28805
